### ISSUE TITLE
Haproxy 2.2.13 too much fd in sock state

### ISSUE THREAD

**Issue by vladsakalkin** [haproxy/haproxy#1231]
## Detailed description of the problem

We have observed increase of the FD files after updating from version 1.9.13 to 2.2.13. 
Before update we had 60K FD opened at peak time. After:  250K. Haproxy stats also shows that we have 250K current connections in Frontend section.
But if I run `ss -t` command, I have only 33K connections in different states (ESTAB, CLOSE_WAIT, etc) and 1.2K etries in 'show sess' command. Then I noticed that we have too much FD allocated for haproxy process, and a lot of them have a sock state:
haproxy 977 haproxy  118u     sock                0,9      0t0   72241549 protocol: TCP
haproxy 977 haproxy  119u     sock                0,9      0t0   72394465 protocol: TCP
haproxy 977 haproxy  120u     sock                0,9      0t0   72946201 protocol: TCP 

## Expected behavior

reasonable FD allocation

## Steps to reproduce the behavior

1. Update to 2.2.13
2. Push requests
The most of our requests via HTTP2

## Do you have any idea what may have caused this?
No(

## Do you have an idea how to solve the issue?
No

## What is your configuration?
```
global
    nbproc 1
    nbthread 30

    cpu-map auto:1/1-30 0-29

    maxconn 524200

    chroot /var/lib/haproxy
    user haproxy
    group haproxy
    daemon

    log 10.21.21.11:10000 len 2048 local1 info
    log-send-hostname haproxy

    stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
    stats timeout 30s
    stats timeout 30s

    ca-base /etc/ssl/private/ca
    crt-base /etc/ssl/private

    ssl-default-bind-ciphers ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:EDH-RSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA:!DSS
    ssl-default-bind-options ssl-min-ver TLSv1.0 no-tls-tickets
    ssl-default-server-ciphers ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:EDH-RSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA:!DSS
    ssl-default-server-options no-sslv3 no-tls-tickets

    tune.http.logurilen 1024
    tune.bufsize 65534
    tune.http.cookielen 4096
    tune.maxrewrite 16384
    tune.ssl.cachesize 1024000
    tune.ssl.default-dh-param 2048

    hard-stop-after 4h
	
defaults
    maxconn 524200

    mode http

    log global
    
    option httplog
    option dontlognull
    option redispatch

    retries 2

    timeout check 10s
    timeout connect 10s
    timeout client  5m
    timeout server  5m
    timeout http-request 1m
    timeout http-keep-alive 2m
    timeout queue 1m

listen ha-statistics
    bind 10.10.10.10:9000
    option forwardfor
    option httpclose
    stats auth admin:admin
    stats enable True
    stats realm Haproxy\ Statistics
    stats refresh 10s
    stats uri /stats
	
frontend service_1
    bind 10.233.18.15:443 ssl crt service_1.pem alpn h2,http/1.1
    capture request header Accept len 4096
    capture request header Accept-Encoding len 4096
    capture request header Accept-Language len 4096
    capture request header Host len 4096
    capture request header Referer len 4096
    capture request header User-Agent len 4096
    capture request header X-Forwarded-For len 4096
    capture request header X-Requested-With len 4096
    capture request header Authorization len 4096
    http-request set-header X-Real-IP %[src]
    http-request set-header X-F5-IP %[src]
    http-request set-header X-Forwarded-For %[src]
    http-request set-header X-Forwarded-Proto https
    log-format 4x|:,_|%B|:,_|%[capture.req.hdr(0)]|:,_|%[capture.req.hdr(1)]|:,_|%[capture.req.hdr(2)]|:,_|-|:,_|%[capture.req.hdr(3)]|:,_|%[capture.req.hdr(4)]|:,_|%[capture.req.hdr(5)]|:,_|%[capture.req.hdr(6)]|:,_|%ci|:,_|%r|:,_|%Ta|:,_|%ST|:,_|%Tl|:,_|%Tr|:,_|%[capture.req.hdr(7)]|:,_|%[capture.req.hdr(8)]|:,_|%cp|:,_|%Ts.%ms|:,_|-|:,_|
    default_backend service_1

backend service_1
    balance roundrobin
    option httpchk
    http-check connect
    http-check send meth GET uri /system/check ver HTTP/1.1 hdr Host service1.domain.org
    http-check expect string OK
    http-response set-header x-edge-processing-time %Tr
    server dc1_backend-1 10.233.9.106:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc2_backend-1 10.233.4.106:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc1_backend-2 10.233.10.132:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc2_backend-2 10.233.5.200:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc1_backend-3 10.233.18.111:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc2_backend-3 10.233.19.111:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc1_backend-4 10.233.18.150:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc2_backend-4 10.233.19.150:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc1_backend-5 10.233.7.207:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc2_backend-5 10.233.2.239:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc1_backend-6 10.233.7.239:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc2_backend-6 10.233.2.240:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc1_backend-7 10.233.2.241:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc2_backend-7 10.233.7.241:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc1_backend-8 10.233.7.252:80 check inter 5s fall 3 rise 3 source 10.233.91.96
    server dc2_backend-8 10.233.2.250:80 check inter 5s fall 3 rise 3 source 10.233.91.96
```

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.2.13-1ppa1~bionic 2021/04/02 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.13.html
Running on: Linux 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-VuoNC3/haproxy-2.2.13=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-stringop-overflow -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=32).
Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.31 2018-02-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 7.5.0
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
	[CACHE] cache
	[FCGI] fcgi-app

Linux 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

## If HAProxy crashed: Last outputs and backtraces

```
-
```

**Comment by wtarreau** [comment]
Just to be sure, is this stable or is this constantly growing ?

Since your client timeouts are quite large (5mn) it could simply be that you're dealing with 800 connections per second and that they stay open for 300s during some transfers, or that you have a bit more and that the 2mn keep-alive timeout maintains them opened. The difference could be that in the past haproxy was less capable of maintaining connections opened (especially before HTX where the two sides were slightly more tightly coupled and errors would more often result in closing the client). Just trying to guess of course.

Could you please post the output of `show info` on such a node, and could you please post the output of `show fd` somewhere ? (beware, it will be large, please check first that it doesn't contain private information).

**Comment by vladsakalkin** [comment]
It's stable, it's rising after restart/reload and then stopped. So first of all I decided that the connections are still being closed, but I'm confused by fd sock state and that in version 1.9 I had another number of conns :) 

show info:
```
Name: HAProxy
Version: 2.2.13-1ppa1~bionic
Release_date: 2021/04/02
Nbthread: 30
Nbproc: 1
Process_num: 1
Pid: 977
Uptime: 0d 8h47m09s
Uptime_sec: 31629
Memmax_MB: 0
PoolAlloc_MB: 1569
PoolUsed_MB: 1316
PoolFailed: 0
Ulimit-n: 1048553
Maxsock: 1048553
Maxconn: 524200
Hard_maxconn: 524200
CurrConns: 212206
CumConns: 47665309
CumReq: 125594069
MaxSslConns: 0
CurrSslConns: 212206
CumSslConns: 47432596
Maxpipes: 0
PipesUsed: 0
PipesFree: 0
ConnRate: 1460
ConnRateLimit: 0
MaxConnRate: 2555
SessRate: 1460
SessRateLimit: 0
MaxSessRate: 2555
SslRate: 1456
SslRateLimit: 0
MaxSslRate: 2553
SslFrontendKeyRate: 1366
SslFrontendMaxKeyRate: 2369
SslFrontendSessionReuse_pct: 6
SslBackendKeyRate: 0
SslBackendMaxKeyRate: 0
SslCacheLookups: 26009798
SslCacheMisses: 23375379
CompressBpsIn: 0
CompressBpsOut: 0
CompressBpsRateLim: 0
ZlibMemUsage: 0
MaxZlibMemUsage: 0
Tasks: 215939
Run_queue: 3
Idle_pct: 92
node: ****
Stopping: 0
Jobs: 212220
Unstoppable Jobs: 0
Listeners: 13
ActivePeers: 0
ConnectedPeers: 0
DroppedLogs: 0
BusyPolling: 0
FailedResolutions: 0
TotalBytesOut: 849344459575
BytesOutRate: 33868608
DebugCommandsIssued: 0
Build info: 2.2.13-1ppa1~bionic
```
haproxy_fd:
https://drive.google.com/file/d/1znYLsl8NFcM-t2JMYtH5MjC--_zdUlrc/view?usp=sharing

**Comment by wtarreau** [comment]
Very interesting. So these are essentially H2 connections, the few H1 ones being end-to-end upgrades (thus most likely websocket):

```
$ fgrep back=0 haproxy_fd | grep -o 'mux=[^ ]*'|sort|uniq -c|sort -n
      1 mux=PASS
   1784 mux=H1
 210295 mux=H2
```

OK there is something wrong. The vast majority of them (203602) are showing that a shutdown read was received, and among them, 198682 without any prior request (i.e. browser preconnect that was not used). I have no idea what could cause this especially since it's a very basic problem, we'll need to be able to reproduce it based on your config, as there's probably a side effect of a combination of settings that plays here.

**Comment by capflam** [comment]
I pushed a patch that should fix this issue. I'll backport it in the 2.3 today. It will be shipped with the next 2.3 release. Backport to the 2.2 is planned for the next week. But in mean time, it may be applied manually on top of the 2.2.

**Comment by capflam** [comment]
The 2.3.10 is out. I'm marking the issue as fixed. But of course, it must be confirmed. And if I'm wrong, please re-open it.

**Comment by capflam** [comment]
The 2.2.14 was released and the fix was backported to the 2.0 too.
